File: src/main/php/ZonedDateTimePersistence.php
<?php

declare(strict_types=1);

namespace PetrKnap\Persistence\ZonedDateTime;

use DateTimeImmutable;
use DateTimeInterface;
use DateTimeZone;
use InvalidArgumentException;

/**
 * @phpstan-import-type LocalDateTime from JavaSe8\Time
 * @phpstan-import-type ZonedDateTime from JavaSe8\Time
 */
final class ZonedDateTimePersistence
{
    private function __construct()
    {
    }

    /**
     * @return ($zonedDateTime is null ? null : LocalDateTime)
     */
    public static function computeUtcDateTime(DateTimeInterface|null $zonedDateTime): DateTimeImmutable|null
    {
        return $zonedDateTime !== null
            ? JavaSe8\Time::toLocalDateTime(JavaSe8\Time::zonedDateTime($zonedDateTime)->setTimezone(new DateTimeZone('UTC')))
            : null;
    }

    /**
     * @note use named arguments for arguments after `$_`, {@see https://www.php.net/manual/en/functions.arguments.php#functions.named-arguments}
     *
     * @param null $_ named arguments separator
     *
     * @return ($utcDateTime is null ? null : ZonedDateTime)
     */
    public static function computeZonedDateTime(
        DateTimeInterface|null $utcDateTime,
        $_ = null,
        DateTimeInterface|null $localDateTime = null,
        DateTimeZone|null $timezone = null,
    ): DateTimeImmutable|null {
        $utcDateTime = JavaSe8\Time::localDateTime($utcDateTime);
        $localDateTime = JavaSe8\Time::localDateTime($localDateTime);

        return match (true) {
            $localDateTime !== null => self::computeZonedDateTimeFromUtcDateTimeAndLocalDateTime($utcDateTime, $localDateTime),
            $timezone !== null => self::computeZonedDateTimeFromUtcDateTimeAndTimezone($utcDateTime, $timezone),
            $utcDateTime === null => null,
            default => throw new InvalidArgumentException('Too few arguments'),
        };
    }

    /**
     * @param LocalDateTime|null $utcDateTime
     * @param LocalDateTime|null $localDateTime
     *
     * @return ZonedDateTime|null
     */
    private static function computeZonedDateTimeFromUtcDateTimeAndLocalDateTime(
        DateTimeImmutable|null $utcDateTime,
        DateTimeImmutable|null $localDateTime,
    ): DateTimeImmutable|null {
        return $utcDateTime !== null && $localDateTime !== null ? DateTimeUtils::asUtcInstantAtOffset(
            $utcDateTime,
            DateTimeUtils::secondsBetween($utcDateTime, $localDateTime),
        ) : null;
    }

    /**
     * @param LocalDateTime|null $utcDateTime
     *
     * @return ZonedDateTime|null
     */
    private static function computeZonedDateTimeFromUtcDateTimeAndTimezone(
        DateTimeImmutable|null $utcDateTime,
        DateTimeZone|null $timezone = null,
    ): DateTimeImmutable|null {
        return $utcDateTime !== null && $timezone !== null ? DateTimeUtils::asUtcInstantAtOffset(
            $utcDateTime,
            0,
        )->setTimezone($timezone) : null;
    }
}

File: src/main/java/io/github/petrknap/persistence/zoneddatetime/ZonedDateTimePersistence.java
package io.github.petrknap.persistence.zoneddatetime;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.time.LocalDateTime;
import java.time.ZoneId;
import java.time.ZonedDateTime;

public final class ZonedDateTimePersistence {
    private ZonedDateTimePersistence() {
    }

    public static @Nullable LocalDateTime computeUtcDateTime(@Nullable ZonedDateTime zonedDateTime) {
        return zonedDateTime != null ? zonedDateTime.withZoneSameInstant(ZoneId.of("UTC")).toLocalDateTime() : null;
    }

    public static @Nullable ZonedDateTime computeZonedDateTime(
            @Nullable LocalDateTime utcDateTime,
            @Nullable LocalDateTime localDateTime
    ) {
        return utcDateTime != null && localDateTime != null ? DateTimeUtils.asUtcInstantAtOffset(
                utcDateTime,
                DateTimeUtils.secondsBetween(utcDateTime, localDateTime)
        ) : null;
    }

    public static @Nullable ZonedDateTime computeZonedDateTime(
            @Nullable CharSequence utcDateTime,
            @Nullable CharSequence localDateTime,
            @NotNull String format
    ) throws Exception.CouldNotComputeZonedDateTime {
        try {
            return utcDateTime != null && localDateTime != null ? computeZonedDateTime(
                    DateTimeUtils.parseAsLocalDateTime(utcDateTime, format),
                    DateTimeUtils.parseAsLocalDateTime(localDateTime, format)
            ) : null;
        } catch (DateTimeUtils.Exception.CouldNotParseAsLocalDateTime cause) {
            throw new Exception.CouldNotComputeZonedDateTime(cause);
        }
    }

    interface Exception {
        final class CouldNotComputeZonedDateTime extends RuntimeException implements Exception {
            public CouldNotComputeZonedDateTime(@NotNull Throwable cause) {
                super(cause.getMessage(), cause);
            }
        }
    }
}
